Zend/tests/enum/019.phpt

```diff
@@ -0,0 +1,14 @@
+--TEST--
+Enum serialize
+--FILE--
+<?php
+
+enum Foo {
+    case Bar;
+}
+
+echo serialize(Foo::Bar);
```

Comment:
The behavior of unserialize should also be tested and documented - it seems like `__unserialize` and `Serializable::unserialize()` and `__wakeup` all would work on brand new instances in the unserializer.

From https://github.com/php/php-src/pull/6489/files#diff-a0e04c509860f6a936aed39d8d35821e9c41f1a5a507570e65465b60691009dd - should I assume a different immutable instance would be created if ReflectionClass->newInstance() is also supported, and that the long term plan is to support isIdentical() on algebraic data types?

Reply:
Yeah, this part is completely missing. I'll need to go through all the various places where new objects are created.

Reply:
At a glance, it seems like it's doable in php 8.1 internals to make the unserializer return a singleton if you override the C `unserialize` callback of the class's `zend_class_entry` (this could introduce edge cases during unserialization failure for destruction of objects "created" by unserialize(), expected to be solvable or not matter since destructors can't be declared).
This would require using the `C:` serialize() encoding (php uses that for classes implementing Serializable), not `o:`
(unserialize can be set to a generic C function that creates a value and puts it in `*object*`)

unserialize() sets `object` to a brand new php value, but I assume it could call ZVAL_OBJ() and possibly increment reference counts to set it to a pre-existing singleton of that class (from the class constant).

```c
// Zend/zend.h
struct _zend_class_entry {
        /* .... serializer callbacks */
        int (*serialize)(zval *object, unsigned char **buffer, size_t *buf_len, zend_serialize_data *data);
        int (*unserialize)(zval *object, zend_class_entry *ce, const unsigned char *buf, size_t buf_len, zend_unserialize_data *data);

// ext/standard/var_unserializer.c, in object_custom()
        if (ce->unserialize == NULL) {
                zend_error(E_WARNING, "Class %s has no unserializer", ZSTR_VAL(ce->name));
                object_init_ex(rval, ce);
        } else if (ce->unserialize(rval, ce, (const unsigned char*)*p, datalen, (zend_unserialize_data *)var_hash) != SUCCESS) {
                return 0;
        }
```

Reply:
> This would require using the C: `serialize()` encoding (php uses that for classes implementing Serializable)

On second thought, using a different letter just for enum types would probably be safer, but the `->serialize` pointer can hopefully still be overridden.

- Serializable has various known limitations such as not being able to share serialized arrays/references/objects - `__serialize()` was added to replace it.
- This allows optimizations for enum types and is more flexible - it permits future extensions to enum types such as properties.
- Unserializing `ClassName::enumName` won't behave as expected in older php versions (PHP_Incomplete_Class?)

Reply:
Done (even though the implementation requires some more work).